### ISSUE TITLE
False positive (I think): eslint-plugin-unicorn(no-useless-spread) `[...new Array(n)]`

### ISSUE THREAD

**Issue by jbr** [oxc-project/oxc#1735]
This is a low confidence report but seemed like it might be helpful feedback.

Currently, `no-useless-spread` wants to replace `[...new Array(n)]` with `new Array(n)`, which is not identical. Spreading a sized array transforms it from `n empty items` into a realized array of that many undefineds. `new Array(2).map((_, n) => n)` is `[<2 empty items>]` and `[...new Array(2)].map((_, n) => n)` is `[0, 1]`

I haven't used eslint-plugin-unicorn before, but [xo](https://github.com/xojs/xo) includes those lints and `npx xo` on a project that uses `[...new Array(n)]` does not emit any no-useless-spread warnings, so it seems like an implementation difference

**Comment by magic-akari** [comment]
Yes, `[...new Array(3)]` and `new Array(3)` are not the same. The former creates an array with three `undefined` elements, while the latter creates an array with three empty holes. However, what surprises me is that it's in the test case.

https://github.com/oxc-project/oxc/blob/2046e8566489ef7012579bc34bc75698e2d817e1/crates/oxc_linter/src/rules/unicorn/no_useless_spread.rs#L561

**Comment by camc314** [comment]
shall we still report this, but instead change the diagnostic method?

in my opinion, i prefer `new Array(n).fill(undefined)`  over `[...new Array(3)]` as it is slightly clearer. it also avoids creating a new array.

```
  ⚠ eslint-plugin-unicorn(no-useless-spread): Using a spread operator here creates a new array unnecessarily.
   ╭─[no_useless_spread.tsx:1:1]
 1 │ [...new Array(3)]
   ·  ───
   ╰────
  help: Using `new Array(n)` creates a new array of length `n`, with holes. If you need the array to be filled with `undefined`, use `new Array(n).fill(undefined)` instead.
  ```

^^ i need to reword the `help` a bit, but you get the idea 🙂

**Comment by Boshen** [comment]
I thought `Array.from({ length:4 })` is the proper way 😅

**Comment by camc314** [comment]
yep, we've got a rule:

https://github.com/oxc-project/oxc/blob/128406a8ab6a2b780c317d20a74693141e77933e/crates/oxc_linter/src/rules/unicorn/no_new_array.rs#L26

If i understand this correctly

```
interface ArrayConstructor {
    new (arrayLength?: number): any[];
    new <T>(arrayLength: number): T[];
    new <T>(...items: T[]): T[];
    (arrayLength?: number): any[];
    <T>(arrayLength: number): T[];
    <T>(...items: T[]): T[];
    isArray(arg: any): arg is any[];
    readonly prototype: any[];
}
```

`new Array(5)` could be different to `new Array(5)`?


Not sure how (a typechecker) would differentiate between `new <T>(...items: T[]): T[];` and `new <T>(arrayLength: number): T[];`

**Comment by Boshen** [comment]
I kind of change the meaning of `correctness` to "identifying erroneous, redundant, or confusing code" (from my announcement).

Does this mean we can change `no_new_array` to correctness, and omit `[...new Array(3)]` from `no_useless_spread`?

**Comment by Boshen** [comment]
See previous discussion: 

https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1994#issuecomment-1328611628
